USDT NS fixes

dalehamel opened the pull request:
Fixes #1071 (ignore #1084 )

This functionality use to work in bpftrace 0.9.2 with BCC up to 0.11.0, It seems to now be broken. Once bcc 0.13.0 is released and bpftrace uses it, it should work again.

Most of the bugs were introduced in bcc, so the only extra handling needed in bpftrace seems to be to check the mount namespace when resolving binaries, like we already do for uprobes.

This also builds off of #1021

davemarchevsky commented:
@dalehamel Happy to help unbreak `bcc` for you if the mount ns changes did end up preventing you from doing something that worked in the past.

dalehamel commented:
thanks @davemarchevsky i think the big thing will be to get a regression test in to bcc using something like `mountns_wrapper` like we do in our runtime tests here. I also want to confirm 100% that the issue is BCC with a minimally reproducible test case (which would probably turn into that regression test).

I'm fairly certain the PR that broke this was https://github.com/iovisor/bcc/pull/2324, as when I tested some weeks ago with an older BCC (0.11.0, prior to that commit), USDT probes worked with an older version of bpftrace that didn't need these fixes here, but a newer version (master / 0.12.0) didn't.

The issue seems specific to USDT probse when in a different mount namespace, as the bpftrace tests for USDT probes within the same ns seem to work just fine.

I'll file an issue against BCC once I'm 100% sure the issue is BCC and not something i've missed in bpftrace, and I'll ping you on it as well as the PR to fix it :)

dalehamel commented:
actually, the root cause may not be BCC at all, the regression tests I've written are passing so far, I think the issue may just be when probing libraries, indicating that #1082 could fix this as well

dalehamel commented:
hmm nope, there is definitely something funky going on here. With the latest bpftrace but using bcc 0.10.0, I can attach to a test process with USDT:

```
bpftrace -e 'usdt::memcached:command__set { printf("HERE\n")}' -p 3978339
```

But the same bpftrace code against 0.11.0 or 0.12.0:

```
Attaching 1 probe...
cannot attach uprobe, probe entry may not exist
Error attaching probe: usdt:memcached:command__set, to PID: 3978339
```

I bisected this further, and found that 01ee0b61a854e06cb62272ddfbb4eeaf16e9ad7c introduces this issue for certain, as its parent commit doesn't have this problem. I found an existing issue for this on the bcc repo https://github.com/iovisor/bcc/issues/2558

so @davemarchevsky it looks like it was your commit that introduced the issue after all, and there is also a *separate* problem in bpftrace related to how it handles paths, which I'll aim to fix here too.

What is weirder even is that my basic mountns_wrapper regression test doesn't catch this :/ so there must be something else going on here.

davemarchevsky commented:
@dalehamel In your repro case, mind peeking at `/proc/PID/maps` of the PID you're trying to attach USDT to? Specifically, are the paths in that file relative to root of that PID's mountns, or the "actual" root `bpftrace` is ostensibly running in?

dalehamel commented:
@davemarchevsky in both cases (5.3 kernel locally, 4.14 kernel in production) it is relative to the PID's mountns:

```
cat /proc/3978339/maps
5617b426b000-5617b426f000 r--p 00000000 08:01 525616                     /usr/local/bin/memcached
5617b426f000-5617b42a5000 r-xp 00004000 08:01 525616                     /usr/local/bin/memcached
5617b42a5000-5617b42b3000 r--p 0003a000 08:01 525616                     /usr/local/bin/memcached
5617b42b3000-5617b42b4000 r--p 00047000 08:01 525616                     /usr/local/bin/memcached
5617b42b4000-5617b42b5000 rw-p 00048000 08:01 525616                     /usr/local/bin/memcached
5617b42b5000-5617b42d0000 rw-p 00000000 00:00 0
5617b4485000-5617b45c1000 rw-p 00000000 00:00 0                          [heap]
...
```

dalehamel commented:
@davemarchevsky here's a clue, I straced the working parent commit and the failing call at https://github.com/iovisor/bcc/blob/071eef6b1571f23be6f0fc03787523cb06b70659/src/cc/libbpf.c#L993 that shows the error message provides a bit of a hint:

The working binary shows the written buffer as:

```
...
write(16, "p:uprobes/p__usr_local_bin_memcached_7a66_1_bcc_293830 /usr/local/bin/memcached:0x7a66", 86) = 86
...
```

The failed one shows:

```
...
write(13, "p:uprobes/p__proc_3978339_root_usr_local_bin_memcached_7a66_1_bcc_296056 /proc/3978339/root/usr/local/bin/memcached:0x7a66", 122) = -1 ENOENT (No such file or directory)
write(2, "cannot attach uprobe, probe entry may not exist\n", 48cannot attach uprobe, probe entry may not exist
...
```

So it looks like the failing one is prepending /proc/PID/root, which may be the cause of the failure?

dalehamel commented:
Digging through the commit, this looks like the culprit https://github.com/iovisor/bcc/commit/01ee0b61a854e06cb62272ddfbb4eeaf16e9ad7c#diff-918ef818c481c73f92c4ce607b1ec386R678 but i'm not sure why

dalehamel commented:
@davemarchevsky what is really strange is that probing this process locally with a 5.3 kernel, this code path doesn't seem to be hit **at all**. I wonder if the issue is that older kernels are hitting a dead code path, which you were never able to test? I assume your development box was probably on a newer kernel...

In any case, the bug seems to be that it is writing a path relative to /proc, when the code is already entering the mount namespace, but this path won't be valid unless the process uses the host process namespace. I set the process to use HostPID in kubernetes and I'm able to probe it, which I think confirms this theory. Without hostpid, the full namespace path is not valid, as the separate process namespace won't have a valid path.

dalehamel commented:
Yeah, it looks like newer kernels probably have `bpf_try_perf_event_open_with_probe` succeed, at https://github.com/iovisor/bcc/blob/071eef6b1571f23be6f0fc03787523cb06b70659/src/cc/libbpf.c#L1025 which explains why this is a regression only on older kernels.

dalehamel commented:
https://github.com/iovisor/bcc/pull/2710 is up to fix the bcc regression, which I think i've fully explained

dalehamel commented:
BCC regression has been fixed, but I discovered another one (with overlayfs). There are a few other bugs here related to listing and attaching USDT probes, which i'll aim to fix in this PR still.

dalehamel commented:
Once https://github.com/iovisor/bcc/pull/2723 is merged, I think that this PR should actually be pretty minimal, as I think most of the regressions were in BCC

dalehamel commented:
Using the latest BCC (master) I am able to finally USDT probe processes inside containers again. We probably won't pick up these fixes until bpftrace 0.9.5 I'm guessing, as we'll need BCC 0.13.0 to be cut...

The actual test regression should ideally test both probing in a separate PID *and* mount namespace, as well as on overlayfs, however this is already tested in BCC with my latest changes there, so I don't see any purpose in doing these tests again in bpftrace.

dalehamel commented:
@fbs @danobi please take a look

dalehamel commented:
oh woops didn't notice the failure @danobi will fix

dalehamel commented:
Ah d'oh, it's because we're still linking against an older bcc that doesn't have my fix @danobi. I'll note this and we can skip the test for now.

I *did* notice another bug though, where we may prepend `/proc/PID/root` multiple times, so I fixed that while I was in here (as I introduced this bug and that's pretty ugly, though it doesn't seem to have broken any behavior yet).

fbs commented:
@dalehamel  what is the status of this? Is it ready to be merged? :)

dalehamel commented:
@fbs yes it should be, the change @danobi requested was for a broken test that I think is because of the BCC version. I've skipped the test for now to get it back to green on travis.

dalehamel commented:
I also fixed that bug I mentioned above

danobi commented on src/utils.cpp:
```diff
@@ -628,9 +628,20 @@ resolve_binary_path(const std::string &cmd, const char *env_paths, int pid)
+
+  snprintf(pid_root, sizeof(pid_root), "/proc/%d/root", pid);
+
+  if (path.find(pid_root) == std::string::npos)
```
Would it be more correct to check this instead?
```suggestion
  if (path.find(pid_root) != 0)
```

dalehamel replied:
yes I think you're right, i'll give this a try

fbs commented:
Look ok, should this be squashed?

dalehamel commented:
Yeah I’ll squash shortly

dalehamel commented:
thanks @fbs
